driver_test.go

```diff
@@ -661,3 +661,366 @@ func TestExecContextDml(t *testing.T) {
 	}
 
 }
+
+func TestExecContextDdl(t *testing.T) {
+
+	// Open db.
+	ctx := context.Background()
+	db, err := sql.Open("spanner", dsn)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	// Run tests.
+	tests := []struct {
```

Comment:
Could we normalise the struct used between tests (from this test and the other PR)? Maybe something like:

```
struct {
  name string
  given []string
  when []string
  then struct {
    query: string
    expectation: struct {
      rows: []row
      error: boolean
    }
  }
  tearDown: []string
}
```

Is this general enough to work for all the tests?

I think if we do this, it will be easier to parse over different PRs.

Reply:
I can change to this kind of data struct for all the tests with ddl/dml.  Though for the query tests I think it's cleaner to just build the table & tear it down at the start and end of the function rather than for every case as it isn't involved in the tests and saves several mins when running on remote database

Reply:
I agree, we shouldn't be creating / dropping the table on every test since it is very time consuming, but the above mentioned struct should be flexible enough to run any queries (ddl / dml) or the given / when / then blocks. So, my question is more related to: could we have the same struct being used for both tests?

Reply:
The same struct couldn't be used across all test functions easily as they all have different checks/ numbers of errors they're looking for. Any foreign key test also need more than one ordered set up & tear down fields. So you'd end up with one giant struct in which half the fields would be ignored in a given test. By both do you mean exec and query tests? For query the given field wouldn't ever be used and without it I just like the classic if (want != got) test loop

Reply:
> The same struct couldn't be used across all test functions easily as they all have different checks/ numbers of errors they're looking for.

Not sure I understand this statement. Wouldn't the example struct that I proposed be general enough for each of the tests (note you'd still have a for loop, defining each test case with one instance of the struct)?

> Any foreign key test also need more than one ordered set up & tear down fields.

This is why the struct proposed has a `given: []string` and a `tearDown: []string`, so that we can execute each statement within the arrays in order. Or maybe I am misunderstanding this?

> So you'd end up with one giant struct in which half the fields would be ignored in a given test.

Are you referring to the example given? If so, I would not think it is giant, but you are correct that sometimes some of the fields will not be used for some tests (like `setup`, `tearDown`, ...). I think, though, that our `struct` should be flexible enough to support all use cases.

> By both do you mean exec and query tests?

I think we should try to use the same struct in all the tests, so that all of them are normalised to the same structure.

> For query the given field wouldn't ever be used and without it I just like the classic if (want != got) test loop

I think that the problem we are trying to solve is to normalise all the tests. If we are able to do so, even if some tests do not use the struct, we could potentially extract the code that gets a list of such structs and performs the checks to be used by all the tests. This way, your test code would only have the specification and would re-use the checking code.

We can also talk IRL if you want to discuss this approach.

Reply:
>Not sure I understand this statement. Wouldn't the example struct that I proposed be general enough for each of the tests (note you'd still have a for loop, defining each test case with one instance of the struct)?

Usually there are errors in a few different places to look for, e.g. you don't might not want an error in the queryContext but you do want an error in rows.Scan, so there'd need to be at least errorExec, errorQuery, errorRowsNext, errorScan, errorRowsClose, errorRowsErr  to cover all the possible errs that I can think of 

>This is why the struct proposed has a given: []string and a tearDown: []string, so that we can execute each statement within the arrays in order. Or maybe I am misunderstanding this?

Right for some reason I missed they were arrays, that would introduce another two loops for every test loop. I can change all tests but I don't love adding more complexity to all the tests to normalise them with the one annoying test if that makes sense

Reply:
> Usually there are errors in a few different places to look for, e.g. you don't might not want an error in the queryContext but you do want an error in rows.Scan, so there'd need to be at least errorExec, errorQuery, errorRowsNext, errorScan, errorRowsClose, errorRowsErr to cover all the possible errs that I can think of

It is good practice to do one assertion per test, so I would say that checking a single error in the struct should be enough.

> Right for some reason I missed they were arrays, that would introduce another two loops for every test loop.

Two loops for every test loop should not add much more time here, since the array size is always very small. 

>  I can change all tests but I don't love adding more complexity to all the tests to normalise them with the one annoying test if that makes sense

Why would you have to change all the tests (other than using the new struct everywhere)? I would propose for us to extract the asserting code into a common place.

Reply:
We decided to park this conversation for now. We can normalise and refactor the tests later.

Reply:
>Why would you have to change all the tests (other than using the new struct everywhere)? I would propose for us to extract the asserting code into a common place.

Ah I misunderstood I thought you meant just normalise each function so they used the same struct and a similar general structure. That'd be a massive change and very different from anything I've come across in go driver tests, I can try and get started on it but it'd take some time getting my head around how it would work.

>Two loops for every test loop should not add much more time here, since the array size is always very small.

Sorry I meant code complexity not algorithmic, which can be disregarded if it's all in one spot

>It is good practice to do one assertion per test, so I would say that checking a single error in the struct should be enough.

If  all the assertions were in the same place, you'd also need to know where you want the error so there'd still need to be some kind of flag to put the assertion on the right error